ReactNativeClient/lib/services/KeymapService.ts

```diff
@@ -0,0 +1,202 @@
+		{ accelerator: 'CommandOrControl+E', command: 'startExternalEditing', platforms: ['darwin', 'win32', 'linux', 'freebsd'] },
+		{ accelerator: 'CommandOrControl+Alt+T', command: 'setTags', platforms: ['darwin', 'win32', 'linux', 'freebsd'] },
+		{ accelerator: 'CommandOrControl+,', command: 'options', platforms: ['linux', 'win32', 'freebsd'] },
+		{ accelerator: 'CommandOrControl+G', command: 'goToAnything', platforms: ['darwin', 'win32', 'linux', 'freebsd'] },
+		{ accelerator: 'F1', command: 'openWebsite', platforms: ['darwin', 'win32', 'linux', 'freebsd'] },
+		// Keymap items with null "accelerator" property allows users to set their own Accelerators
+		// i.e. { accelerator: null, command: 'someCommand', platforms: ['darwin', 'win32', 'linux', 'freebsd'] }
+	];
+
+	constructor() {
+		super();
+
+		this.keymap = KeymapService.initKeymap();
+		this.logger_ = new Logger();
+	}
+
+	getAccelerator(command: string) {
+		const item = this.keymap[command];
+		// It's okay to provide a null Accelerator for Electron menu items by default
+		return (item == null) ? null : item.accelerator;
+	}
+
+	setAccelerator(command: string, accelerator: string) {
+		this.keymap[command].accelerator = accelerator;
+	}
+
+	resetAccelerator(command: string) {
+		const defaultItem = KeymapService.defaultKeymap.find((item => item.command === command));
+		this.setAccelerator(command, defaultItem.accelerator);
+	}
+
+	getKeymap() {
+		const defaultItemsByCommand = KeymapService.initKeymap();
+
+		// Compare each "accelerator" value of the current keymap with the default values
+		// Return a keymap object, which can later be converted to a string and saved to the disk
+		return Object.entries(this.keymap).reduce((customKeymap, [command, { accelerator }]) => {
+			// If the Accelerator has been changed, add into the customKeymap object
+			if (defaultItemsByCommand[command].accelerator !== accelerator) {
+				customKeymap.push({ command, accelerator });
+			}
+			return customKeymap;
+		}, []);
+	}
+
+	setKeymap(customKeymap: KeymapItem[]) {
+		this.logger().info('Updating keymap...');
+
+		return new Promise((resolve, reject) => {
+			for (let i = 0; i < customKeymap.length; i++) {
+				const item = customKeymap[i];
+
+				// Validation errors will be caught and logged
+				if (!item.hasOwnProperty('command')) {
+					this.logger().warn(`Command is missing in: ${JSON.stringify(item)}`);
+				} else if (!item.hasOwnProperty('accelerator')) {
+					this.logger().warn(`Accelerator is missing in: ${JSON.stringify(item)}`);
+				} else if (!this.keymap.hasOwnProperty(item.command)) {
+					this.logger().warn(`Command is invalid in: ${JSON.stringify(item)}`);
+				} else if (!(item.accelerator === null || KeymapService.isAccelerator(item.accelerator))) {
+					this.logger().warn(`Accelerator is invalid in: ${JSON.stringify(item)}`);
+				} else {
+					this.setAccelerator(item.command, item.accelerator);
+				}
```

Comment:
Please create a separate function to validate a keymap.

What does the user do when a keymap is not valid? How do we tell them?